src/array.rs

```diff
@@ -0,0 +1,456 @@
+    }
+
+    /// New array from a bool scalar.
+    pub fn from_bool(val: bool) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_bool(val) };
+        Array { ptr }
+    }
+
+    /// New array from a int scalar.
+    pub fn from_int(val: i32) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_int(val) };
+        Array { ptr }
+    }
+
+    /// New array from a float scalar.
+    pub fn from_float(val: f32) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_float(val) };
+        Array { ptr }
+    }
+
+    /// New array from a complex scalar.
+    pub fn from_complex(val: c64) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_complex(val.re, val.im) };
+        Array { ptr }
+    }
+
+    /// New array from existing buffer.
+    ///
+    /// # Parameters
+    ///
+    /// - `data`: A buffer which will be copied.
+    /// - `shape`: Shape of the array.
+    /// - `dim`: Number of dimensions (size of shape).
+    pub fn from_data<T: ArrayElement>(data: &[T], shape: &[i32], dim: i32) -> Self {
+        let ptr = unsafe {
+            mlx_sys::mlx_array_from_data(
+                data.as_ptr() as *const c_void,
+                shape.as_ptr(),
+                dim,
+                T::DTYPE as u32,
+            )
+        };
+
+        Array { ptr }
+    }
+
+    /// The size of the array’s datatype in bytes.
+    pub fn itemsize(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_itemsize(self.ptr) }
+    }
+
+    /// Number of elements in the array.
+    pub fn size(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_size(self.ptr) }
+    }
+
+    /// The strides of the array.
+    pub fn strides(&self) -> &[usize] {
+        let ndim = self.ndim();
+
+        unsafe {
+            let data = mlx_sys::mlx_array_strides(self.ptr);
+            std::slice::from_raw_parts(data, ndim)
+        }
+    }
+
+    /// The number of bytes in the array.
+    pub fn nbytes(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_nbytes(self.ptr) }
+    }
+
+    /// The array’s dimension.
+    pub fn ndim(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_ndim(self.ptr) }
+    }
+
+    /// The shape of the array.
+    ///
+    /// Returns: a pointer to the sizes of each dimension.
+    pub fn shape(&self) -> &[i32] {
+        let ndim = self.ndim();
+
+        unsafe {
+            let data = mlx_sys::mlx_array_shape(self.ptr);
+            std::slice::from_raw_parts(data, ndim)
+        }
+    }
+
+    /// The shape of the array in a particular dimension.
+    ///
+    /// # Panic
+    ///
+    /// Panics if the array is scalar.
+    pub fn dim(&self, dim: i32) -> i32 {
+        // This will panic on a scalar array
+        unsafe { mlx_sys::mlx_array_dim(self.ptr, dim) }
+    }
+
+    /// The array element type.
+    pub fn dtype(&self) -> Dtype {
+        let dtype = unsafe { mlx_sys::mlx_array_get_dtype(self.ptr) };
+        Dtype::try_from(dtype).unwrap()
+    }
+
+    /// Evaluate the array.
```

Comment:
might leave a comment here about how MLX is lazy and arrays are not fully realized until they are evaluated.

Reply:
I feel like that should go to the ops, but we could make it a TODO after we implement the ops